### PR TITLE
[WIP] [Do not merge] Reduce overhead of AOT Module

### PR DESCRIPTION
**Pull request by anijain2305**
@jansel @Chillee 

Inspired heavily from - https://github.com/facebookresearch/torchdynamo/blob/ce0c84a62d3287e2afde22e0d823a8d1ae4758a8/torchdynamo/optimizations/python_key.py#L121

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -506,24 +555,119 @@ def aot_module(mod: nn.Module, *args, **kwargs) -> nn.Module:
...
+    #########################################################
+    """
+    (1) Create a new fn_for_tracing that lifts params as inputs (TODO: buffers)
+    (2) A new tracer - MyTracer (slight modification of PythonKeyTracer). This
+    works with Proxy tensors instead of PythonTensors. Goal is to get a torch
+    graph, and not torch.aten graph yet.
+    (3) This traced function is then passed on to aot_function.
+    (4) The params are read and flattened on every forward call, as they can change during training.
+    """
+    #########################################################
```
Comment (anijain2305) [inline functorch/_src/aot_autograd.py:568]: @jansel @Chillee This is the flow.

File: functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
 from .named_members_polyfill import _named_parameters, _named_buffers
 from typing import Callable, List, Dict, Any, Tuple, Optional
 
+import torchdynamo
```
Comment (jansel) [inline functorch/_src/aot_autograd.py:15]: perhaps:
```
try:
    from torchdynamo import disable as disable_dynamo
except ImportError:
    def disable_dynamo(x):
         return x
```

So we can still use this without dynamo.

File: functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -506,24 +555,119 @@ def aot_module(mod: nn.Module, *args, **kwargs) -> nn.Module:
...
+        traced = torch.fx.GraphModule(tracer.root, graph, "my_tracer")
+        traced.recompile()
+
+
+        def fn_with_params_as_args(*joint_args):
+            return traced(*joint_args)
+        return fn_with_params_as_args
+
+
+    gm = torch.fx.symbolic_trace(mod)
```
Comment (jansel) [inline functorch/_src/aot_autograd.py:641]: Why do we need this?  Seems not ideal, as this could be lossy.

File: functorch/_src/aot_autograd.py [hunk 4]
```diff
@@ -506,24 +555,119 @@ def aot_module(mod: nn.Module, *args, **kwargs) -> nn.Module:
...
+            params = {
+                **dict(_named_parameters(gm, remove_duplicate=False)),
+                **dict(_named_buffers(gm, remove_duplicate=False)),
+            }
+            if HAS_TREE:
+                params_flat = tree.flatten(params)
+            else:
+                params_flat, _ = pytree.tree_flatten(params)
+
+            params_flat = tuple(params_flat)
```
Comment (jansel) [inline functorch/_src/aot_autograd.py:662]: In the torchdynamo case (not in the general case), this list is a constant you don't need to recompute every time.

File: functorch/_src/aot_autograd.py [hunk 5]
```diff
@@ -506,24 +555,119 @@ def aot_module(mod: nn.Module, *args, **kwargs) -> nn.Module:
...
+            joint_args = (*params_flat, *args) 
+            nargs = len(joint_args)
+
+            if compiled_f is None:
+                fn_with_params_as_args = flattened_fn(gm, nargs)
+                compiled_f = aot_function(fn_with_params_as_args, *top_args, **top_kwargs)
```
Comment (jansel) [inline functorch/_src/aot_autograd.py:668]: In the torchdynamo case (not in the general case), there should only every be one `compiled_f`.  You can just compute it ahead of time and don't need to wait until the first call.

File: functorch/_src/aot_autograd.py (hunk 4 above)
Comment (anijain2305) [reply functorch/_src/aot_autograd.py:662]: For training, the params will change after each update. Don't we have to flatten for each call?

File: functorch/_src/aot_autograd.py (hunk 5 above)
Comment (anijain2305) [reply functorch/_src/aot_autograd.py:668]: Ok. Yeah, we can send example_inputs and compile it beforehand.

Comment (jansel) [reply functorch/_src/aot_autograd.py:668]: I think there is also a bunch of pytree overhead inside aot_function, plus an unneeded caching layer.

File: functorch/_src/aot_autograd.py (hunk 4 above)
Comment (anijain2305) [reply functorch/_src/aot_autograd.py:662]: Yeah, I realized while driving today. These are just references. And the update happens in place.

Will remove it.

**Comment by anijain2305** [comment]
Closing. Following PRs are covering this

* Contiguous - https://github.com/pytorch/functorch/pull/536
* AOT module simplified - https://github.com/pytorch/functorch/pull/660
* Disabling torchdynamo - https://github.com/pytorch/functorch/pull/662

**Pull request closed without merging**